validators/cgroup_validator_linux.go

```diff
@@ -44,28 +44,117 @@ func (c *CgroupsValidator) Name() string {
 }
 
 const (
-	cgroupsConfigPrefix = "CGROUPS_"
-	unifiedMountpoint   = "/sys/fs/cgroup"
+	cgroupsConfigPrefix      = "CGROUPS_"
+	defaultUnifiedMountpoint = "/sys/fs/cgroup"
+	mountsFilePath           = "/proc/mounts"
+	mountInfoPath            = "/proc/self/mountinfo"
 )
 
+// getUnifiedMountpoint checks if the default mount point is available.
+// If not, it parses the mounts file to find a valid cgroup mount point.
+func getUnifiedMountpoint() (string, error, error) {
+	isCgroupfs, err := isMountPointCgroupfs(defaultUnifiedMountpoint)
+	if isCgroupfs {
+		return defaultUnifiedMountpoint, err, nil
+	}
+	mountPath, err := parseMountsFile(mountsFilePath)
+	return mountPath, err, fmt.Errorf("%q is not cgroupfs: %v", defaultUnifiedMountpoint, err)
+}
+
+func isMountPointCgroupfs(path string) (bool, error) {
+	f, err := os.Open(mountInfoPath)
+	if err != nil {
+		return false, err
+	}
+
+	defer f.Close()
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if !strings.Contains(line, "cgroup") {
+			continue
+		}
+		// example fields: `28 22 0:25 / /sys/fs/cgroup rw,nosuid,nodev,noexec,relatime shared:4 - cgroup2 cgroup2 rw,seclabel`
+		fields := strings.Fields(line)
+		if len(fields) >= 9 {
+			if fields[5] == path {
+				i := 6
+				for i < len(fields) && fields[i] != "-" {
+					i++
+				}
+				if len(fields)-i <= 3 {
+					return false, fmt.Errorf("expect 3 fields after `-` in mount info file %q", path)
+				}
+				return strings.Contains(fields[i+2], "cgroup"), nil
+			}
+		}
+	}
+	return false, fmt.Errorf("path %q is not cgroupfs", path)
+}
+
+func parseMountsFile(path string) (string, error) {
```

Comment:
aren't we doing the same thing two times?

- checking a mount point path with isMountPointCgroupfs
- then reading the same file again with this function parseMountsFile

i think this can be done in one go.

here we don't care much about performance but it's just better code org.

Reply:
```
###  /proc/mounts is the same as  /proc/self/mounts
[root@paco ~]# cat /proc/mounts   | grep cgroup
cgroup2 /sys/fs/cgroup cgroup2 rw,seclabel,nosuid,nodev,noexec,relatime 0 0
none /run/cilium/cgroupv2 cgroup2 rw,seclabel,relatime 0 0
[root@paco ~]# cat /proc/self/mountinfo   | grep cgroup
28 22 0:25 / /sys/fs/cgroup rw,nosuid,nodev,noexec,relatime shared:4 - cgroup2 cgroup2 rw,seclabel
1499 27 0:25 / /run/cilium/cgroupv2 rw,relatime shared:573 - cgroup2 none rw,seclabel
```

Probably, we can just parce the ` /proc/mounts` for the first cgroup v2 path or first cgroup v1 path.